clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -373,15 +374,53 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         }
         // Invariant: non-explicit control flow (i.e., kernel-mediated) is indicated
         // by markers.
+        bool have_cond_branch_target = false;
+        addr_t cond_branch_target = 0;
         if (shard->prev_instr_.instr.addr != 0 /*first*/ &&
-            !type_is_instr_branch(shard->prev_instr_.instr.type)) {
+            type_is_instr_direct_branch(shard->prev_instr_.instr.type) &&
+            // We do not bother to support legacy traces without encodings.
+            TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_)) {
+            addr_t trace_pc = shard->prev_instr_.instr.addr;
+            if (shard->prev_instr_.instr.encoding_is_new)
+                shard->branch_target_cache.erase(trace_pc);
+            auto cached = shard->branch_target_cache.find(trace_pc);
+            if (cached != shard->branch_target_cache.end()) {
+                have_cond_branch_target = true;
+                cond_branch_target = cached->second;
+            } else {
+                instr_t instr;
+                instr_init(GLOBAL_DCONTEXT, &instr);
+                const app_pc decode_pc =
+                    const_cast<app_pc>(shard->prev_instr_.instr.encoding);
+                const app_pc next_pc =
+                    decode_from_copy(GLOBAL_DCONTEXT, decode_pc,
+                                     reinterpret_cast<app_pc>(trace_pc), &instr);
+                if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
```

Comment:
I would lump that in with an invalid encoding too which shouldn't happen: the next_pc==nullptr.  I was thinking of this as an invariant check that the encoding is actually a valid branch, catching errors in raw2trace inserting encoding records or in the reader properly placing encoding data.  It's a gray area: leaning toward an invariant error rather than an assert.